apps/fortuna/src/command/register_provider.rs

```diff
@@ -28,42 +37,65 @@ pub struct CommitmentMetadata {
 /// Register as a randomness provider. This method will generate and commit to a new random
 /// hash chain from the configured secret & a newly generated random value.
 pub async fn register_provider(opts: &RegisterProviderOptions) -> Result<()> {
-    let chain_config = Config::load(&opts.config.config)?.get_chain_config(&opts.chain_id)?;
+    let config = Config::load(&opts.config.config)?;
+    let chain_config = config.get_chain_config(&opts.chain_id)?;
+
+    register_provider_from_config(&config.provider, &opts.chain_id, &chain_config).await?;
+
+    Ok(())
+}
+
+pub async fn register_provider_from_config(
+    provider_config: &ProviderConfig,
+    chain_id: &ChainId,
+    chain_config: &EthereumConfig,
+) -> Result<()> {
+    let private_key_string = provider_config
+        .private_key
+        .load()?
+        .ok_or(anyhow!(
+            "Please specify a provider private key in the config"
+        ))?
+        .clone();
 
     // Initialize a Provider to interface with the EVM contract.
     let contract =
-        Arc::new(SignablePythContract::from_config(&chain_config, &opts.private_key).await?);
+        Arc::new(SignablePythContract::from_config(&chain_config, &private_key_string).await?);
     // Create a new random hash chain.
     let random = rand::random::<[u8; 32]>();
-    let secret = opts.randomness.load_secret()?;
+    let secret = provider_config
+        .secret
+        .load()?
+        .ok_or(anyhow!("Please specify a provider secret in the config"))?;
 
-    let commitment_length = opts.randomness.chain_length;
+    let commitment_length = provider_config.chain_length;
     let mut chain = PebbleHashChain::from_config(
         &secret,
-        &opts.chain_id,
-        &opts.private_key.clone().parse::<LocalWallet>()?.address(),
+        &chain_id,
+        &private_key_string.parse::<LocalWallet>()?.address(),
         &chain_config.contract_addr,
         &random,
         commitment_length,
     )?;
 
     // Arguments to the contract to register our new provider.
-    let fee_in_wei = opts.fee;
+    let fee_in_wei = chain_config.fee;
     let commitment = chain.reveal()?;
     // Store the random seed and chain length in the metadata field so that we can regenerate the hash
     // chain at-will. (This is secure because you can't generate the chain unless you also have the secret)
     let commitment_metadata = CommitmentMetadata {
         seed:         random,
         chain_length: commitment_length,
     };
+    let uri = get_register_uri(&provider_config.uri, &chain_id)?;
```

Comment:
drive-by bugfix: the previous implementation registered different URLs if you used register-provider vs. setup-provider